service/matching/taskListManager_test.go

```diff
@@ -388,3 +388,85 @@ func TestGetPollerIsolationGroup(t *testing.T) {
 	assert.Equal(t, 1, len(groups))
 	assert.Equal(t, config.AllIsolationGroups[0], groups[0])
 }
+
+// The intent of this test: SingleTaskDispatch is one of two places where tasks are written to
+// the taskreader.taskBuffers channels. As such, it needs to take care to not accidentally
+// hit the channel when it's full, as it'll block, causing a deadlock (due to both this dispatch
+// and the async task pump blocking trying to read to the channel)
+func TestSingleTaskDispatchDoesNotDeadlock(t *testing.T) {
+	controller := gomock.NewController(t)
+
+	config := defaultTestConfig()
+	config.EnableTasklistIsolation = func(domain string) bool { return true }
+	tlm := createTestTaskListManagerWithConfig(controller, config)
+
+	// mock a timeout to cause tasks to be rerouted
+	tlm.taskReader.dispatchTask = func(ctx context.Context, task *InternalTask) error {
+		return context.DeadlineExceeded
+	}
+	tlm.taskReader.getIsolationGroupForTask = func(ctx context.Context, info *persistence.TaskInfo) (string, error) {
+		// force this, on the second read, to always reroute to the other DC
+		return "datacenterB", nil
+	}
+
+	maxBufferSize := config.GetTasksBatchSize("", "", 0) - 1
+
+	tlm.taskReader.logger = loggerimpl.NewNopLogger()
+
+	for i := 0; i < maxBufferSize; i++ {
+		breakDispatcher, breakRetryLoop := tlm.taskReader.dispatchSingleTaskFromBuffer("datacenterA", &persistence.TaskInfo{})
+		assert.False(t, breakDispatcher, "dispatch isn't shutting down")
+		assert.True(t, breakRetryLoop, "should be able to successfully dispatch all these tasks to the other isolation group")
+	}
+	// We should see them all being redirected
+	assert.Len(t, tlm.taskReader.taskBuffers["datacenterB"], maxBufferSize)
+	// we should see none in DC A (they're not being redirected here)
+	assert.Len(t, tlm.taskReader.taskBuffers["datacenterA"], 0)
+
+	// ok, and here we try and ensure that this *does not block
+	// and instead complains and live-retries
+	breakDispatcher, breakRetryLoop := tlm.taskReader.dispatchSingleTaskFromBuffer("datacenterA", &persistence.TaskInfo{})
+	assert.False(t, breakDispatcher, "dispatch isn't shutting down")
+	assert.False(t, breakRetryLoop, "in the event of a task being unable to be cross-dispatched, the expectation is that it'll keep retrying")
```

Comment:
```suggestion
	// ok, and here we try and ensure that this *does not block
	// and instead complains and live-retries
	breakDispatcher, breakRetryLoop := tlm.taskReader.dispatchSingleTaskFromBuffer("datacenterA", &persistence.TaskInfo{})
	assert.False(t, breakDispatcher, "dispatch isn't shutting down")
	assert.False(t, breakRetryLoop, "in the event of a task being unable to be cross-dispatched, the expectation is that it'll keep retrying")
	
	// and test again, to cover max-buffer +1, to make sure "keep retrying"
	// is a safe and stable state to be in.
	breakDispatcher, breakRetryLoop := tlm.taskReader.dispatchSingleTaskFromBuffer("datacenterA", &persistence.TaskInfo{})
	assert.False(t, breakDispatcher, "dispatch isn't shutting down")
	assert.False(t, breakRetryLoop, "in the event of a task being unable to be cross-dispatched, the expectation is that it'll keep retrying")
```

Reply:
hmm.  or maybe I misinterpreted.  just seems odd that max-minus-1 is actually "full", and it doesn't occur when _exceeding_ max.

Reply:
good call